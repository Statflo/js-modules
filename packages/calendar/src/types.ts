export enum CalendarServices {
    GOOGLE = 'google',
    ICS = 'ics',
    ICAL = 'ical',
    APPLE = 'apple',
    OUTLOOK = 'outlook',
};

export type CalendarServiceTypes =
    | CalendarServices.GOOGLE
    | CalendarServices.ICS
    | CalendarServices.ICAL
    | CalendarServices.APPLE
    | CalendarServices.OUTLOOK;

export type DateArgument = Date | string;

export interface CalendarEventConfig {
    start: DateArgument;
    end?: DateArgument;
    title?: string;
    location?: string;
    description?: string;
    url?: string;
}

export interface CreateCalendarEvent extends CalendarEventConfig {
    service: CalendarServiceTypes;
}
