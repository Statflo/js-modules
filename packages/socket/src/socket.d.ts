import * as Stomp from '@stomp/stompjs/esm5';
import { SocketServiceSubscriptions, SocketServiceMessageCallback, SocketServiceOptions } from './types';
export declare class SocketService {
    client: Stomp.Client;
    connected?: boolean;
    subscriptions: SocketServiceSubscriptions;
    private static readonly defaultOptions;
    private static readonly defaultHeaders;
    constructor(url: string, connectHeaders?: Stomp.StompHeaders, options?: SocketServiceOptions);
    subscribe(target: string, callback: SocketServiceMessageCallback): void;
    unsubscribe(target: string): void;
    connect(): Promise<Stomp.Frame | undefined>;
    disconnect(): Promise<{}>;
    message(destination: string, body: string): void;
}
export default SocketService;
