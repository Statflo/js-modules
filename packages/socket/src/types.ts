import * as SockJS from 'sockjs-client';
import * as Stomp from '@stomp/stompjs';

export interface SocketServiceOptions {
    sockJS?: SockJS.Options;
}

export type SocketServiceMessageCallback = (message: Stomp.Message) => void;

export type SocketServiceSubscriptions = Map<string, Stomp.StompSubscription>;

export declare class SocketService {
    client: Stomp.Client;
    connected?: boolean;
    subscriptions: SocketServiceSubscriptions;
    private static readonly defaultOptions;
    private static readonly defaultHeaders;
    constructor(url: string, connectHeaders?: Stomp.StompHeaders, options?: SocketServiceOptions);
    subscribe(target: string, callback: SocketServiceMessageCallback): Stomp.StompSubscription;
    unsubscribe(target: string): void;
    connect(): Promise<Stomp.Frame | undefined>;
    disconnect(): Promise<{}>;
    message(destination: string, body: string): void;
}

export default SocketService;
