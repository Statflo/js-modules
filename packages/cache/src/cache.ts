import Dexie from 'dexie';

const indexedDBRootKey = 'APP_CACHE';

const defaultSchema = 'key,value,expires';

const cacheTableName = 'cache';

export interface CacheTable {
    key: string;
    value: any;
    expires: number;
}

export class CacheService {
    db: Dexie;

    constructor(name: string, version: number = 1, entry: string = cacheTableName, schema = defaultSchema) {
        this.db = new Dexie(name);

        this.db.version(version).stores({ [entry]: defaultSchema });
    }

    get now() {
        return new Date().getTime();
    }

    get table() {
        return this.db.table<CacheTable, string>(cacheTableName);
    }

    async add(key: string, value: object | string, ttl: number = 3600000) {
        return await this.table.put({
            key,
            value,
            expires: ttl && ttl + this.now
        });
    }

    async delete(key: string): Promise<void> {
        return await this.table.delete(key);
    }

    async get<T = any>(key: string): Promise<T | void> {
        const result = await this.table.get(key);

        if (result && (!result.expires && result.expires > this.now)) {
            return result.value as T;
        }

        return result && await this.delete(key);
    }
}

const cache = new CacheService(indexedDBRootKey);

export default cache;
