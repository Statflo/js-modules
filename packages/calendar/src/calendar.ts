import googleService from './services/google';
import icsService from './services/ics';
import {
    CalendarServices,
    CreateCalendarEvent
} from './types';

export const google = googleService;
export const ics = icsService;

export const services = {
    [CalendarServices.GOOGLE]: google,
    [CalendarServices.ICS]: ics,
    [CalendarServices.ICAL]: ics,
    [CalendarServices.APPLE]: ics,
    [CalendarServices.OUTLOOK]: ics
};

export function createCalendarEvent({ service, ...config }: CreateCalendarEvent): string {
    return services[service](config);
};

export default createCalendarEvent;
