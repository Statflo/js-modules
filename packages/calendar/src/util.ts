import { DateArgument, CalendarEventConfig } from './types';

const timePattern = /-|:|\.\d+/g;
const checkISODateEligibility = (date: DateArgument): string => date instanceof Date ? date.toISOString() : date;
const formatTime = (date: DateArgument): string => checkISODateEligibility(date).replace(timePattern, '');

export const getDefaultConfig = (config: CalendarEventConfig): CalendarEventConfig => ({
    title: '',
    location: '',
    description: '',
    url: '',
    ...config,
    start: config.start ? formatTime(config.start) : '',
    end: config.end ? formatTime(config.end) : '',
});
