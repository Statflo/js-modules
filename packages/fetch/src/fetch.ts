export type FetchToolsInput = Request | string;

export interface FetchToolsCache<CacheShape> {
    [url: string]: CacheShape;
}

export type FetchMiddleware<P = any> = (input: FetchToolsInput, init?: RequestInit, previousResult?: P | void) => Promise<P>;

export type FetchMiddlewareStack<P = any> = FetchMiddleware<P>[];

export interface FetchMiddlewareHybridType extends Function {
    add?: (middleware: FetchMiddleware) => FetchMiddlewareHybridType;
}

const cacheResponse = <CacheShape>(cache: FetchToolsCache<CacheShape>, endPoint: string) =>
    (response: CacheShape) => Promise.resolve((cache[endPoint] = response));

const parseJSON = (response: Response) => response.json();

export function checkStatus(response: Response) {
    if (response.status >= 200 && response.status < 300) {
        return response;
    } else {
        const error = new Error(response.statusText);
        throw error;
    }
};

export function fetchJSON<JSONShape = object>(input: Request | string, init?: RequestInit) {
    return fetch(input, init).then(checkStatus).then<JSONShape>(parseJSON);
}

export function createCacheableFetch<CacheShape = object>(cache: FetchToolsCache<CacheShape> = Object.create(null)) {
    return function fetchAndCache(url: string, init?: RequestInit) {
        return cache[url] && Promise.resolve(cache[url])
            || fetchJSON<CacheShape>(url, init).then(cacheResponse<CacheShape>(cache, url));
    }
}

export function createMiddleware<ResultShape = any>(stack: FetchMiddlewareStack<ResultShape> = [], fetchHandler: Function = fetch) {
    const runMiddlewareStack = <FetchMiddlewareHybridType>async function fetchWithMiddleware(input: FetchToolsInput, init?: RequestInit) {
        let previousResult;

        for (let middleware of stack) {
            previousResult = await middleware(input, init, previousResult);
        }

        return fetchHandler(input, init);
    };

    runMiddlewareStack.add = function addMiddleware(middleware) {
        return (stack = stack.concat(middleware)) && runMiddlewareStack;
    }

    return runMiddlewareStack;
}
