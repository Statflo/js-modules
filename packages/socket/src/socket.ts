import * as SockJS from 'sockjs-client';
import {
    Stomp,
    CompatClient,
    Message,
    Frame,
    StompHeaders
} from '@stomp/stompjs/esm5';

export interface SocketServiceOptions {
    sockJS?: SockJS.Options;
}

export type SocketServiceMessageCallback = (message: Message) => void;

export class SocketService {
    client: CompatClient;
    connected?: boolean;

    private static readonly defaultOptions = {
        sockJS: {}
    }

    public constructor(url: string, options: SocketServiceOptions = SocketService.defaultOptions) {
        console.log('url', url);
        this.client = Stomp.over(new SockJS(url, null, options.sockJS));
        console.log('this', this);
    }

    public subscribe(target: string, callback: SocketServiceMessageCallback) {
        return this.client.subscribe(target, callback);
    }

    public connect(headers: StompHeaders): Promise<Frame | undefined> {
        return new Promise((resolve, reject) => this.client.connect(headers, (frame: Frame) => {
            this.connected = true;
            resolve(frame);
        }, (msg: Error) => {
            this.connected = false;
            reject(msg);
        }));
    }

    public disconnect() {
        return new Promise(resolve => {
            this.connected = false;
            this.client.disconnect(resolve)
        });
    }

    public message(target: string, body: string) {
        if (!this.connected) {
            throw new Error('Socket client service is not connected');
        }

        this.client.send(target, {}, body);
    }
}

export default SocketService;
