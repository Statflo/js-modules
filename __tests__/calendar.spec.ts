import createCalendarEvent, { google, ics, services } from '@statflo/calendar';
import { CalendarEventConfig, CalendarServices, CalendarServiceTypes } from '@statflo/calendar/src/types';

enum TestTopic {
    GOOGLE = 'Google Calendar',
    ICS = 'Apple Calendar and Outlook'
}

const requiredConfig: CalendarEventConfig =  {
    start: '2018-05-10T02:00:26.578Z',
};

const requiredConfigDate: CalendarEventConfig = {
    start: new Date(requiredConfig.start)
};

const fullConfig: CalendarEventConfig = {
    ...requiredConfig,
    end: '2018-05-10T02:09:08.138Z',
    title: 'Foobar',
    location: '1 Yonge St, Toronto - ON - Canada',
    description: 'Monthly potluck',
    url: 'https://www.statflo.com'
};

const fullConfigDate: CalendarEventConfig = {
    ...fullConfig,
    ...requiredConfigDate,
    end: new Date(fullConfig.end)
};


describe('@statflo/calendar', () => {

    describe('ISO string dates', () => {
        describe(TestTopic.GOOGLE, () => {
            it('should run required config', () => {
                 expect(google(requiredConfig)).toMatchSnapshot();
            });

            it('should run full config', () => {
                expect(google(fullConfig)).toMatchSnapshot();
            });
        });

        describe(TestTopic.ICS, () => {
            it('should run required config', () => {
                expect(ics(requiredConfig)).toMatchSnapshot();
            });

            it('should run full config', () => {
                expect(ics(fullConfig)).toMatchSnapshot();
            });
        });
    });

    describe('Date class', () => {
        describe(TestTopic.GOOGLE, () => {
            it('should run required config', () => {
                expect(google(requiredConfigDate)).toMatchSnapshot();
            });

            it('should run full config', () => {
                expect(google(fullConfigDate)).toMatchSnapshot();
            });
        });

        describe(TestTopic.ICS, () => {
            it('should run required config', () => {
                expect(ics(requiredConfigDate)).toMatchSnapshot();
            });

            it('should run full config', () => {
                expect(ics(fullConfigDate)).toMatchSnapshot();
            });
        });
    });

    describe('Module integration', () => {
        it('Should follow module shape', () => {
            expect(google).toBeInstanceOf(Function);
            expect(ics).toBeInstanceOf(Function);
            expect(createCalendarEvent).toBeInstanceOf(Function);

            for (let service in CalendarServices) {
                expect(services[CalendarServices[service]]).toBeInstanceOf(Function);
            }
        });

        it('Should run createCalendarEvent with all service scenarios', () => {
            for (let service in CalendarServices) {
                expect(createCalendarEvent({
                    service: CalendarServices[service] as CalendarServiceTypes,
                    start: requiredConfig.start
                })).toBe(services[CalendarServices[service]]({
                    start: requiredConfig.start
                }));
            }
        });
    });
});
