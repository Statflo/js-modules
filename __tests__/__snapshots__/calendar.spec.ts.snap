// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@statflo/calendar Date class Apple Calendar and Outlook should run full config 1`] = `"data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:https://www.statflo.com%0ADTSTART:20180510T020026Z%0ADTEND:20180510T020908Z%0ASUMMARY:Foobar%0ADESCRIPTION:Monthly%20potluck%0ALOCATION:1%20Yonge%20St,%20Toronto%20-%20ON%20-%20Canada%0AEND:VEVENT%0AEND:VCALENDAR"`;

exports[`@statflo/calendar Date class Apple Calendar and Outlook should run required config 1`] = `"data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:%0ADTSTART:20180510T020026Z%0ADTEND:%0ASUMMARY:%0ADESCRIPTION:%0ALOCATION:%0AEND:VEVENT%0AEND:VCALENDAR"`;

exports[`@statflo/calendar Date class Google Calendar should run full config 1`] = `"https://www.google.com/calendar/render?action=TEMPLATE&text=Foobar&dates=20180510T020026Z/20180510T020908Z&details=Monthly%20potluck&location=1%20Yonge%20St,%20Toronto%20-%20ON%20-%20Canada&sprop=&sprop=name:"`;

exports[`@statflo/calendar Date class Google Calendar should run required config 1`] = `"https://www.google.com/calendar/render?action=TEMPLATE&text=&dates=20180510T020026Z/&details=&location=&sprop=&sprop=name:"`;

exports[`@statflo/calendar ISO string dates Apple Calendar and Outlook should run full config 1`] = `"data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:https://www.statflo.com%0ADTSTART:20180510T020026Z%0ADTEND:20180510T020908Z%0ASUMMARY:Foobar%0ADESCRIPTION:Monthly%20potluck%0ALOCATION:1%20Yonge%20St,%20Toronto%20-%20ON%20-%20Canada%0AEND:VEVENT%0AEND:VCALENDAR"`;

exports[`@statflo/calendar ISO string dates Apple Calendar and Outlook should run required config 1`] = `"data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:%0ADTSTART:20180510T020026Z%0ADTEND:%0ASUMMARY:%0ADESCRIPTION:%0ALOCATION:%0AEND:VEVENT%0AEND:VCALENDAR"`;

exports[`@statflo/calendar ISO string dates Google Calendar should run full config 1`] = `"https://www.google.com/calendar/render?action=TEMPLATE&text=Foobar&dates=20180510T020026Z/20180510T020908Z&details=Monthly%20potluck&location=1%20Yonge%20St,%20Toronto%20-%20ON%20-%20Canada&sprop=&sprop=name:"`;

exports[`@statflo/calendar ISO string dates Google Calendar should run required config 1`] = `"https://www.google.com/calendar/render?action=TEMPLATE&text=&dates=20180510T020026Z/&details=&location=&sprop=&sprop=name:"`;
